File: tools/migrate_vault.py
#!/usr/bin/env python3
"""
Vault Migration Tool for ActiveMirrorOS

Migrates vaults from old fixed-salt format to new random-salt format.

Usage:
    python tools/migrate_vault.py <vault_path> <password>

What it does:
1. Detects old vault format (fixed salt)
2. Exports all vault data with old decryption
3. Creates new vault with random salt
4. Re-encrypts all data
5. Preserves metadata and timestamps

Author: AMOS Dev Twin
"""

import sys
import json
import base64
from pathlib import Path
from datetime import datetime
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC


class OldVaultMemory:
    """Old vault implementation with fixed salt - for migration only."""

    FIXED_SALT = b'activemirror_vault'

    def __init__(self, vault_path: str, password: str):
        self.vault_path = Path(vault_path)
        self.password = password

        # Use old key derivation with fixed salt
        kdf = PBKDF2HMAC(
            algorithm=hashes.SHA256(),
            length=32,
            salt=self.FIXED_SALT,
            iterations=100000,
        )
        key = base64.urlsafe_b64encode(kdf.derive(password.encode()))
        self.cipher = Fernet(key)

        self.index_file = self.vault_path / ".vault_index.enc"

    def can_decrypt(self) -> bool:
        """Check if we can decrypt with fixed salt."""
        if not self.index_file.exists():
            return False

        try:
            encrypted_data = self.index_file.read_bytes()

            # Old format: just encrypted data, no salt prefix
            decrypted = self.cipher.decrypt(encrypted_data)
            json.loads(decrypted.decode())
            return True
        except Exception:
            return False

    def export_all(self) -> dict:
        """Export all vault data."""
        if not self.index_file.exists():
            return {"entries": [], "metadata": {}}

        # Load index
        encrypted_index = self.index_file.read_bytes()
        decrypted_index = self.cipher.decrypt(encrypted_index)
        index = json.loads(decrypted_index.decode())

        # Export all entries
        exported_entries = []

        for key, info in index.get("entries", {}).items():
            entry_file = Path(info["file"])

            if entry_file.exists():
                encrypted_entry = entry_file.read_bytes()
                decrypted_entry = self.cipher.decrypt(encrypted_entry)
                entry_data = json.loads(decrypted_entry.decode())

                exported_entries.append({
                    "key": key,
                    "value": entry_data["value"],
                    "metadata": entry_data.get("metadata", {}),
                    "created_at": entry_data.get("created_at"),
                    "updated_at": entry_data.get("updated_at"),
                })

        return {
            "entries": exported_entries,
            "metadata": {
                "vault_created_at": index.get("created_at"),
                "exported_at": datetime.now().isoformat(),
                "entry_count": len(exported_entries),
            }
        }


def migrate_vault(vault_path: str, password: str, backup: bool = True) -> dict:
    """
    Migrate vault from old to new format.

    Args:
        vault_path: Path to vault directory
        password: Vault password
        backup: Whether to backup old vault (default: True)

    Returns:
        Migration report dict
    """
    vault_path = Path(vault_path)

    print(f"🔍 Analyzing vault at: {vault_path}")

    # Check if vault exists
    if not vault_path.exists():
        return {
            "success": False,
            "error": "Vault path does not exist",
        }

    # Try old format first
    old_vault = OldVaultMemory(str(vault_path), password)

    if not old_vault.can_decrypt():
        # Check if it's already new format
        index_file = vault_path / ".vault_index.enc"
        if index_file.exists():
            raw_data = index_file.read_bytes()
            if len(raw_data) >= 16:
                print("✅ Vault is already in new format (has salt prefix)")
                return {
                    "success": True,
                    "already_migrated": True,
                    "message": "Vault already uses random salt format",
                }

        return {
            "success": False,
            "error": "Cannot decrypt vault with old or new format - wrong password?",
        }

    print("📤 Exporting data from old vault format...")
    export_data = old_vault.export_all()

    entry_count = len(export_data["entries"])
    print(f"   Found {entry_count} entries to migrate")

    if entry_count == 0:
        print("⚠️  Vault is empty - nothing to migrate")
        return {
            "success": True,
            "entries_migrated": 0,
            "message": "Vault is empty",
        }

    # Backup old vault
    if backup:
        backup_path = vault_path.parent / f"{vault_path.name}.backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        print(f"💾 Creating backup at: {backup_path}")

        import shutil
        shutil.copytree(vault_path, backup_path)
        print(f"   ✅ Backup created")

    # Save export to JSON (for safety)
    export_file = vault_path.parent / f"{vault_path.name}_export_{datetime.now().strftime('%Y%m%d_%H%M%S')}.json"
    export_file.write_text(json.dumps(export_data, indent=2))
    print(f"💾 Export saved to: {export_file}")

    # Import new VaultMemory (with random salt)
    print("🔄 Importing new VaultMemory...")
    sys.path.insert(0, str(Path(__file__).parent.parent / "sdk" / "python"))
    from activemirror.vault_memory import VaultMemory

    # Delete old vault files
    print("🗑️  Removing old vault files...")
    for file in vault_path.glob("*"):
        file.unlink()

    # Create new vault with random salt
    print("🔐 Creating new vault with random salt...")
    new_vault = VaultMemory(vault_path=str(vault_path), password=password)

    # Import all entries
    print("📥 Importing entries into new vault...")
    migrated_count = 0
    failed_entries = []

    for entry in export_data["entries"]:
        try:
            new_vault.store(
                key=entry["key"],
                value=entry["value"],
                metadata=entry.get("metadata", {}),
            )
            migrated_count += 1
            print(f"   ✅ Migrated: {entry['key']}")
        except Exception as e:
            failed_entries.append({
                "key": entry["key"],
                "error": str(e),
            })
            print(f"   ❌ Failed: {entry['key']} - {e}")

    print(f"\n✨ Migration complete!")
    print(f"   Migrated: {migrated_count}/{entry_count} entries")

    if failed_entries:
        print(f"   ⚠️  Failed: {len(failed_entries)} entries (see export file)")

    return {
        "success": True,
        "entries_migrated": migrated_count,
        "entries_total": entry_count,
        "failed_entries": failed_entries,
        "backup_path": str(backup_path) if backup else None,
        "export_file": str(export_file),
    }


def main():
    """CLI entry point."""
    if len(sys.argv) < 3:
        print("Usage: python tools/migrate_vault.py <vault_path> <password>")
        print("\nExample:")
        print("  python tools/migrate_vault.py ./my_vault my_password")
        print("\nOptions:")
        print("  --no-backup    Skip creating backup")
        sys.exit(1)

    vault_path = sys.argv[1]
    password = sys.argv[2]
    backup = "--no-backup" not in sys.argv

    print("=" * 60)
    print("🔧 ActiveMirrorOS Vault Migration Tool")
    print("=" * 60)
    print()

    result = migrate_vault(vault_path, password, backup=backup)

    print()
    print("=" * 60)

    if result["success"]:
        print("✅ MIGRATION SUCCESSFUL")
        if result.get("already_migrated"):
            print("   Vault already in new format - no changes needed")
        else:
            print(f"   Migrated: {result['entries_migrated']} entries")
            if result.get("backup_path"):
                print(f"   Backup: {result['backup_path']}")
            print(f"   Export: {result['export_file']}")
    else:
        print("❌ MIGRATION FAILED")
        print(f"   Error: {result['error']}")
        sys.exit(1)

    print("=" * 60)


if __name__ == "__main__":
    main()

File: tools/migrate_vault.js
#!/usr/bin/env node

/**
 * Vault Migration Tool for ActiveMirrorOS (JavaScript)
 *
 * Migrates vaults from old fixed-salt format to new random-salt format.
 *
 * Usage:
 *   node tools/migrate_vault.js <vault_path> <password>
 *
 * Author: AMOS Dev Twin
 */

import fs from 'fs/promises';
import { existsSync } from 'fs';
import path from 'path';
import crypto from 'crypto';

class OldVaultMemory {
  static FIXED_SALT = 'activemirror_vault';

  constructor(vaultPath, password) {
    this.vaultPath = vaultPath;
    this.password = password;
    this.algorithm = 'aes-256-gcm';

    // Use old key derivation with fixed salt
    this.encryptionKey = crypto.pbkdf2Sync(
      password,
      OldVaultMemory.FIXED_SALT,
      100000,
      32,
      'sha256'
    );
  }

  async canDecrypt() {
    const indexPath = path.join(this.vaultPath, '.vault_index.enc');

    if (!existsSync(indexPath)) {
      return false;
    }

    try {
      const encryptedData = await fs.readFile(indexPath);

      // Old format: just encrypted data, no salt prefix
      const decrypted = this._decrypt(encryptedData);
      JSON.parse(decrypted);
      return true;
    } catch (error) {
      return false;
    }
  }

  _decrypt(buffer) {
    const iv = buffer.slice(0, 16);
    const authTag = buffer.slice(16, 32);
    const encrypted = buffer.slice(32);

    const decipher = crypto.createDecipheriv(this.algorithm, this.encryptionKey, iv);
    decipher.setAuthTag(authTag);

    let decrypted = decipher.update(encrypted);
    decrypted = Buffer.concat([decrypted, decipher.final()]);

    return decrypted.toString('utf8');
  }

  async exportAll() {
    const indexPath = path.join(this.vaultPath, '.vault_index.enc');

    if (!existsSync(indexPath)) {
      return { entries: [], metadata: {} };
    }

    // Load index
    const encryptedIndex = await fs.readFile(indexPath);
    const decryptedIndex = this._decrypt(encryptedIndex);
    const index = JSON.parse(decryptedIndex);

    // Export all entries
    const exportedEntries = [];

    for (const [key, info] of Object.entries(index.entries || {})) {
      const entryFile = info.file;

      if (existsSync(entryFile)) {
        const encryptedEntry = await fs.readFile(entryFile);
        const decryptedEntry = this._decrypt(encryptedEntry);
        const entryData = JSON.parse(decryptedEntry);

        exportedEntries.push({
          key,
          value: entryData.value,
          metadata: entryData.metadata || {},
          createdAt: entryData.createdAt,
          updatedAt: entryData.updatedAt,
        });
      }
    }

    return {
      entries: exportedEntries,
      metadata: {
        vaultCreatedAt: index.createdAt,
        exportedAt: new Date().toISOString(),
        entryCount: exportedEntries.length,
      },
    };
  }
}

async function migrateVault(vaultPath, password, backup = true) {
  console.log(`🔍 Analyzing vault at: ${vaultPath}`);

  // Check if vault exists
  if (!existsSync(vaultPath)) {
    return {
      success: false,
      error: 'Vault path does not exist',
    };
  }

  // Try old format first
  const oldVault = new OldVaultMemory(vaultPath, password);

  if (!(await oldVault.canDecrypt())) {
    // Check if it's already new format
    const indexPath = path.join(vaultPath, '.vault_index.enc');
    if (existsSync(indexPath)) {
      const rawData = await fs.readFile(indexPath);
      if (rawData.length >= 16) {
        console.log('✅ Vault is already in new format (has salt prefix)');
        return {
          success: true,
          alreadyMigrated: true,
          message: 'Vault already uses random salt format',
        };
      }
    }

    return {
      success: false,
      error: 'Cannot decrypt vault with old or new format - wrong password?',
    };
  }

  console.log('📤 Exporting data from old vault format...');
  const exportData = await oldVault.exportAll();

  const entryCount = exportData.entries.length;
  console.log(`   Found ${entryCount} entries to migrate`);

  if (entryCount === 0) {
    console.log('⚠️  Vault is empty - nothing to migrate');
    return {
      success: true,
      entriesMigrated: 0,
      message: 'Vault is empty',
    };
  }

  // Backup old vault
  let backupPath;
  if (backup) {
    const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
    backupPath = `${vaultPath}.backup_${timestamp}`;
    console.log(`💾 Creating backup at: ${backupPath}`);

    await fs.cp(vaultPath, backupPath, { recursive: true });
    console.log('   ✅ Backup created');
  }

  // Save export to JSON
  const timestamp = new Date().toISOString().replace(/[:.]/g, '-').slice(0, 19);
  const exportFile = path.join(
    path.dirname(vaultPath),
    `${path.basename(vaultPath)}_export_${timestamp}.json`
  );
  await fs.writeFile(exportFile, JSON.stringify(exportData, null, 2));
  console.log(`💾 Export saved to: ${exportFile}`);

  // Import new VaultMemory (with random salt)
  console.log('🔄 Importing new VaultMemory...');
  const sdkPath = path.join(path.dirname(new URL(import.meta.url).pathname), '../sdk/javascript');
  const { VaultMemory } = await import(path.join(sdkPath, 'vault.js'));

  // Delete old vault files
  console.log('🗑️  Removing old vault files...');
  const files = await fs.readdir(vaultPath);
  for (const file of files) {
    await fs.unlink(path.join(vaultPath, file));
  }

  // Create new vault with random salt
  console.log('🔐 Creating new vault with random salt...');
  const newVault = new VaultMemory({ vaultPath, password });
  await newVault.initialize();

  // Import all entries
  console.log('📥 Importing entries into new vault...');
  let migratedCount = 0;
  const failedEntries = [];

  for (const entry of exportData.entries) {
    try {
      await newVault.store(entry.key, entry.value, entry.metadata || {});
      migratedCount++;
      console.log(`   ✅ Migrated: ${entry.key}`);
    } catch (error) {
      failedEntries.push({
        key: entry.key,
        error: error.message,
      });
      console.log(`   ❌ Failed: ${entry.key} - ${error.message}`);
    }
  }

  console.log(`\n✨ Migration complete!`);
  console.log(`   Migrated: ${migratedCount}/${entryCount} entries`);

  if (failedEntries.length > 0) {
    console.log(`   ⚠️  Failed: ${failedEntries.length} entries (see export file)`);
  }

  return {
    success: true,
    entriesMigrated: migratedCount,
    entriesTotal: entryCount,
    failedEntries,
    backupPath,
    exportFile,
  };
}

async function main() {
  const args = process.argv.slice(2);

  if (args.length < 2) {
    console.log('Usage: node tools/migrate_vault.js <vault_path> <password>');
    console.log('\nExample:');
    console.log('  node tools/migrate_vault.js ./my_vault my_password');
    console.log('\nOptions:');
    console.log('  --no-backup    Skip creating backup');
    process.exit(1);
  }

  const vaultPath = args[0];
  const password = args[1];
  const backup = !args.includes('--no-backup');

  console.log('='.repeat(60));
  console.log('🔧 ActiveMirrorOS Vault Migration Tool');
  console.log('='.repeat(60));
  console.log();

  const result = await migrateVault(vaultPath, password, backup);

  console.log();
  console.log('='.repeat(60));

  if (result.success) {
    console.log('✅ MIGRATION SUCCESSFUL');
    if (result.alreadyMigrated) {
      console.log('   Vault already in new format - no changes needed');
    } else {
      console.log(`   Migrated: ${result.entriesMigrated} entries`);
      if (result.backupPath) {
        console.log(`   Backup: ${result.backupPath}`);
      }
      console.log(`   Export: ${result.exportFile}`);
    }
  } else {
    console.log('❌ MIGRATION FAILED');
    console.log(`   Error: ${result.error}`);
    process.exit(1);
  }

  console.log('='.repeat(60));
}

main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
